liferay-forms/liferay-portal
# LPD-34654 & LPD-35736 Handle undo/redo in Date picker and Select From List field

LPD-34654 & LPD-35736 Handle undo/redo in Date picker and Select From List field

# What is this trying to solve?

1.  When using undo/redo, we cannot go back to the "Choose an option" option, we should be able to restore this option when undoing the selection
2.  In Date picker, we cannot go back to the next or previous option when the current state is empty

[Link to ticket Date Picker](https://liferay.atlassian.net/browse/LPD-35736)  
[Link to ticket Select From List Field](https://liferay.atlassian.net/browse/LPD-34654)

# How am I fixing it?

**1.1** [LPD-34654 This is not an array](https://github.com/liferay-forms/liferay-portal/pull/2602/commits/84711dbea2ac4c1c9565eef780604aed268a0ba7): This is just a fix, `selectedItem` is always a string as [defined here](https://github.com/liferay-forms/liferay-portal/blob/84711dbea2ac4c1c9565eef780604aed268a0ba7/modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-field-type/src/main/resources/META-INF/resources/Select/Select.tsx#L73), this fixes an accessibility issue adding the title to the option  

![](https://github.com/user-attachments/assets/e173ccf3-fc80-49e4-b39f-281bc96d8ddd)

**1.2** [LPD-34654 Verify if newValue doesn't exist](https://github.com/liferay-forms/liferay-portal/pull/2602/commits/6f299f5780295100dd2707f5d162469e6850a5a2): We should verify if `newValue` does not exist instead of if it exists

**2.** [LPD-35736 Always update formattedDate](https://github.com/liferay-forms/liferay-portal/pull/2602/commits/e7bf363bf972d434618412adbafadd1b8da7e255): Always update `formattedDate` , because when we have an empty option selexted, we are not able to restore the previous or the next option as it is overwritten.

# How can I test it?

*   The steps are in the ticket descriptions  
*   **Select From List**

1.  Enable FFs LPD-11228 and LPD-15596
2.  Go to Content & Data > Structures
3.  Add a Selec From List field > Add some options
4.  Go to Content & Data > Web Content
5.  Fill the title
6.  Choose an option from the select from list
7.  Undo the action
8.  See that the undo restores restore the “Choose an option“ option

*   **Date Picker**

1.  Go to Content & Data > Structures
2.  Add a Date picker field
3.  Go to Content & Data > Web Content
4.  Change the date picker value
5.  Click on Undo
6.  Click on Redo
7.  See Date picker restores the value

https://github.com/user-attachments/assets/f2de62e7-a8f4-4fc7-9c8b-11c0ef180494

# Notes

**The tests will be send in ahoter task as it's part of Web Content.**

**There is a bug when we change the locale and go back to default locale using undo/redo, the title and description fields are lost, this will be fixed in another task part of Web Content.**

**Thanks!**
